### DISCUSSION TITLE
Unexplained errors 500 with Symfony app - env prod only

### DISCUSSION THREAD

**Discussion by MartG31** [symfony/symfony#46816] [Q&A]
Hi community !

I manage a web application working on Symfony 5.9.3 and i have unusual and unexplained errors on server since my last update.

My website is based on multiple JavaScript ajax requests sent to the server to refresh my dynamic interfaces using json.

Example JS function :

```
this.linkAssocVitrage = function(data) {

	$.ajax({
        url: "{{ path('ministore-suivi-prod-link-assoc') }}",
        type: 'post',
        data: data,
        dataType: 'json',

        beforeSend: function() {
        	console.log('=== ajax linkAssocVitrage launched ===');
        	console.log(data);
        },
        success: function(json, success, status) {
        	console.log('=== ajax linkAssocVitrage success ===');
            console.log(json);

            if(json.success) {
            	ctn = $('body').find('.assoc_vitrage_ctn');
            	module_fsp.buildAssocVitrage(ctn, json.ms);
            }

        },
        error: function(response, status, error) {
        	console.log('=== ajax linkAssocVitrage error ===');
        	console.log(response);
        	console.log(status);
        	console.log(error);
        }
    });
}
```

Example PHP/Symfony json response :
```

/**
 * @Security("is_granted('ROLE_MINISTORE')")
 * @Route("/ministore/suivi-production/link-assoc", name="ministore-suivi-prod-link-assoc")
*/
public function linkAssocVitrage(Request $request) {

    $ms = $this->manager->getRepository(Ministore::class)->find($request->request->get('ms'));
    $ms_assoc = $this->manager->getRepository(MinistoreAssociationVitrage::class)->findOneBy(['ministore' => $ms]);

    $ms_to_link = $this->manager->getRepository(Ministore::class)->find($request->request->get('ms_to_link'));

    if($ms_assoc == null) {
        $maxNumGroup = $this->manager->getRepository(MinistoreAssociationVitrage::class)->getMaxNumGroup();

    	$assoc1 = new MinistoreAssociationVitrage();
    	$assoc1->setMinistore($ms);
        $assoc1->setNumGroup($maxNumGroup+1);
        $assoc1->setPosition(1);

    	$assoc2 = new MinistoreAssociationVitrage();
    	$assoc2->setMinistore($ms_to_link);
        $assoc2->setNumGroup($maxNumGroup+1);
        $assoc2->setPosition(2);

        $this->manager->persist($assoc1);
        $this->manager->persist($assoc2);
        $this->manager->flush();

        $success = true;
    }

    return new JsonResponse([
    	'success' => $success ?? false,
    	'ms' => $this->manager->getRepository(Ministore::class)->formatMinistore($ms),
    ]);
}
```

Principle is very easy and i've got hundreds requests of this kind working on my website. I developed new features yesterday, as usual everything works fine in dev environment. In prod environment, the new ajax requests (only) gives errors 500.

![image](https://user-images.githubusercontent.com/31081479/176682724-b3da1f23-7673-4c93-9077-17a92785a71e.png)

- Unbelievable, each time I reset the server cache with command `php bin/console cache:clear --env=prod`, each of my requests works perfectly once, then the error 500 comes back at next attempts...

- When i force dev mode on the prod server to log it, the errors disappear and everything works fine.

The rest of website keep working perfectly, these new errors concern some of the new features only.
It's hard to understand if these errors come from Symfony env prod configuration or if it's caused by the hosted server (Infomaniak).

If someone has reflection elements to fix the problem, thanks a lot :). I can provide any other information about my environment.

Thanks in advance for your help !
Best regards,
MG

**Comment by lyrixx** [comment]
You should open `var/log/prod.log`, you may find useful information

**Reply by MartG31** [reply]
Hi lyrixx,
Thanks for your suggestion, i agree these logs are often useful. Unfortunately in that case trigger the error 500 doesn't create any log :/.
Regards

**Reply by lyrixx** [reply]
That's not normal. Some hints:

* A bad configuration, and logs are elsewhere
* The directory is not write able
* It's a segfault, or something really bad. So logs are in PHP (or FPM) log system

**Reply by MartG31** [reply]
Okay, i found the way to write these logs by editing my **config/packages/prod/monolog.yaml** passing "action_level" from "error" to "debug" : 

```
monolog:
    handlers:
        main:
            type: fingers_crossed
            action_level: debug
            handler: nested
            excluded_http_codes: [404, 405]
            buffer_size: 50 # How many messages should be saved? Prevent memory leaks
        nested:
            type: stream
            path: "%kernel.logs_dir%/%kernel.environment%.log"
            level: debug
        console:
            type: console
            process_psr_3_messages: false
            channels: ["!event", "!doctrine"]
```

Here is the error log :

> [2022-06-30T18:58:11.654958+02:00] php.INFO: User Deprecated: Since symfony/security-core 5.3: Injecting the "session" in "Symfony\Component\Security\Core\Authentication\Token\Storage\UsageTrackingTokenStorage" is deprecated, inject the "request_stack" instead. {"exception":"[object] (ErrorException(code: 0): User Deprecated: Since symfony/security-core 5.3: Injecting the \"session\" in \"Symfony\\Component\\Security\\Core\\Authentication\\Token\\Storage\\UsageTrackingTokenStorage\" is deprecated, inject the \"request_stack\" instead. at /home/clients/577932a905e1221745865f72af3e3434/sites/app.immoblade.fr/vendor/symfony/security-core/Authentication/Token/Storage/UsageTrackingTokenStorage.php:84)"} []

It's clearly time to upgrade my Symfony version and resolve my depreciations. But I don't explain the direct link between this depreciation and the request concerned, and why all website requests are not impacted in that case...

Any ideas ? Anyway thanks lyrixx for your hints ;)

**Reply by lyrixx** [reply]
This is a deprecation, not the real error. You have to find something at ERROR ou CRITICAL level

**Comment by ghost** [comment]
Please look at the Chrome Developer Tools at Networks tab - here we have most important information. Then look at Nginx and FPM logs, Then application logs in /var/logs.
However I saw something awkward in your JS code. It looks like you send data along with `application/x-www-form-urlencoded` (default) header but expects the server to return json. This is by design? I think what happens is that your data is json but you forgot to add `Content-type: application/json` header when sending request. This results in HTTP 500 because this call `$request->request->get('ms')` will fail.
